lib/spack/spack/binary_distribution.py

```diff
@@ -100,438 +80,368 @@ class BinaryCacheIndex(object):
+                        pass
+
+                # Fast path: someone else updated the index in the moment when we weren't
+                # holding a lock. Our expected_hash may be out-of-date now, so assume
+                # the other guy did good instead of aggressively checking.
+                if mirror_url in metadata:
+                    current_hash = metadata[mirror_url]["hash"]
+                    if current_hash != previous_hash:
+                        return False
+
+                # At this point: we need to fetch the index.json. No two ways about it.
+                # Do the fetch now, so that if it fails we don't leave any cruft behind.
+                # We will catch the exception this raises outside the write_transaction
+                _, _, fs = web_util.read_from_url(index_fetch_url)
+                data = codecs.getreader("utf-8")(fs).read()
+
+                # Decide on the cache key for the data blob, if one hasn't been chosen
+                if mirror_url in metadata:
+                    data_key = metadata[mirror_url]["key"]
+                else:
+                    data_key = newdata_key
+                    if len(metadata) > 0:
+                        # If someone used that key already, use a small discriminator
+                        if any(
+                            data_key == other_mirror["key"] for other_mirror in metadata.values()
+                        ):
+                            _, data_key = self._cache_keys_for(mirror_url, len(metadata))
+                        # If there's corruption and *that* was used already, count up from a large
+                        # random number. We're bound to find something somewhere we can use.
+                        if any(
+                            data_key == other_mirror["key"] for other_mirror in metadata.values()
+                        ):
+                            other_keys = {
+                                other_mirror["key"] for other_mirror in metadata.values()
+                            }
+                            discriminator = random.randrange(2**8 + len(metadata), 2**16)
+                            while data_key in other_keys:
+                                discriminator += 1
+                                _, data_key = self._cache_keys_for(mirror_url, discriminator)
+
+                # Stash the final updated index in the cache at the appropriate key
+                with self._index_file_cache.write_transaction(data_key) as (_, data_f):
+                    data_f.write(data)
+
+                # Update the metadata to match our new reality
+                mdata = metadata.setdefault(mirror_url, {})
+                mdata["key"] = data_key
+                mdata["hash"] = compute_hash(data)
+                sjson.dump(metadata, newmeta_f)
+        except (URLError, web_util.SpackWebError) as e:
+            tty.warn("Ignoring error while fetching binary cache indices", e)
```

Comment:
This is worse error handling than it used to be. Earlier the idea was to collect errors and show them only when they were an actual issue.

Reply:
Note that *any* error here is not an actual issue, since we do not require that the buildcache have an `index.json` (direct fetches are allowed). The new implementation absorbs all errors silently, which makes no user-facing difference if a non-buildcache mirror (e.g. https://mirror.spack.io) was enabled.